File: clients/php/Model/SoundProperties.php
<?php

namespace Model {
    require_once 'Stream.php';

    /**
     * Sound properties
     */
    class SoundProperties
    {
        /**
         * Name
         */
        public string $name;
        /**
         * Distance from which the sound can be heard
         */
        public float $distance;
        /**
         * Offset modifier
         */
        public float $offset;
    
        function __construct(string $name, float $distance, float $offset)
        {
            $this->name = $name;
            $this->distance = $distance;
            $this->offset = $offset;
        }
    
        /**
         * Read SoundProperties from input stream
         */
        public static function readFrom(\InputStream $stream): SoundProperties
        {
            $name = $stream->readString();
            $distance = $stream->readDouble();
            $offset = $stream->readDouble();
            return new SoundProperties($name, $distance, $offset);
        }
        
        /**
         * Write SoundProperties to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeString($this->name);
            $stream->writeDouble($this->distance);
            $stream->writeDouble($this->offset);
        }
    }
}

File: clients/cpp/model/SoundProperties.cpp
#include "SoundProperties.hpp"

namespace model {

SoundProperties::SoundProperties(std::string name, double distance, double offset) : name(name), distance(distance), offset(offset) { }

// Read SoundProperties from input stream
SoundProperties SoundProperties::readFrom(InputStream& stream) {
    std::string name = stream.readString();
    double distance = stream.readDouble();
    double offset = stream.readDouble();
    return SoundProperties(name, distance, offset);
}

// Write SoundProperties to output stream
void SoundProperties::writeTo(OutputStream& stream) const {
    stream.write(name);
    stream.write(distance);
    stream.write(offset);
}

// Get string representation of SoundProperties
std::string SoundProperties::toString() const {
    std::stringstream ss;
    ss << "SoundProperties { ";
    ss << "name: ";
    ss << '"' << name << '"';
    ss << ", ";
    ss << "distance: ";
    ss << distance;
    ss << ", ";
    ss << "offset: ";
    ss << offset;
    ss << " }";
    return ss.str();
}

}